### PR TITLE
Feature/improve and test floating points in exception message

### PR DESCRIPTION
**Pull request by mgovers**
Improve error message with floating point values

Before:

```txt
Iteration failed to converge after 20 iterations! Max deviation: 0.000000, error tolerance: 0.000000.

Try validate_input_data() or validate_batch_data() to validate your data.
```

After:

```txt
Iteration failed to converge after 20 iterations! Max deviation: 3.54512e-16, error tolerance: 1e-20.

Try validate_input_data() or validate_batch_data() to validate your data.
```

@figueroa1395 since I reran the notebook, you will run into some merge conflicts. On your branch, just make sure you're on the latest PGM core version. To resolve the merge conflict, just run `git add docs/examples && git reset docs/examples && git checkout docs/examples` and re-run all of them

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp [hunk 1]
```diff
@@ -8,9 +8,18 @@
 #include "enum.hpp"
 
 #include <exception>
+#include <sstream>
 #include <string>
 
 namespace power_grid_model {
+namespace detail {
+inline auto to_string(std::floating_point auto x) {
+    std::ostringstream sstr{};
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:17]:
```suggestion
    std::ostringstream const sstr{};
```

File: tests/c_api_tests/test_serialization.cpp [hunk 2]
```diff
@@ -138,6 +138,29 @@ TEST_CASE("Serialization") {
             CHECK(is_nan(node[0].u_rated));
             CHECK(source[0].id == 6);
             CHECK(source[1].id == 7);
+
```
Comment (nitbharambe) [inline tests/c_api_tests/test_serialization.cpp:141]: What is the motivation for these additions in serialization?

File: power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp [hunk 3]
```diff
@@ -8,9 +8,18 @@
 #include "enum.hpp"
 
 #include <exception>
+#include <sstream>
 #include <string>
 
 namespace power_grid_model {
+namespace detail {
+inline auto to_string(std::floating_point auto x) {
+    std::ostringstream const sstr{};
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:17]:
```suggestion
    std::ostringstream sstr{};  // NOLINT(misc-const-correctness) // https://github.com/llvm/llvm-project/issues/57297
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp [hunk 4]
```diff
@@ -8,9 +8,18 @@
 #include "enum.hpp"
 
 #include <exception>
+#include <sstream>
 #include <string>
 
 namespace power_grid_model {
+namespace detail {
+inline auto to_string(std::floating_point auto x) {
+    std::ostringstream sstr{};  // NOLINT(misc-const-correctness) // https://github.com/llvm/llvm-project/issues/57297
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:17]:
```suggestion
    // https://github.com/llvm/llvm-project/issues/57297
    std::ostringstream sstr{};  // NOLINT(misc-const-correctness)
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp [hunk 5]
```diff
@@ -8,9 +8,19 @@
...
 
 #include <exception>
+#include <sstream>
 #include <string>
 
 namespace power_grid_model {
+namespace detail {
+inline auto to_string(std::floating_point auto x) {
+    // https://github.com/llvm/llvm-project/issues/57297
+    std::ostringstream sstr{};  // NOLINT(misc-const-correctness)
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/common/exception.hpp:18]:
```suggestion
    std::ostringstream sstr{}; // NOLINT(misc-const-correctness) // https://github.com/llvm/llvm-project/issues/57297
```

File: tests/c_api_tests/test_serialization.cpp (hunk 2 above)
Comment (mgovers) [reply tests/c_api_tests/test_serialization.cpp:141]: whoops sorry unrelated to this PR. this is something I discussed with Laurynas and that i will probably add in another PR (there's no example for using serialization result in the C API). accidentally added to this one. Will revert

File: tests/c_api_tests/test_serialization.cpp [hunk 6]
```diff
@@ -138,6 +138,29 @@ TEST_CASE("Serialization") {
...
+            // create output dataset
+            std::vector<SymNodeOutput> sym_node_output(1);
+            MutableDatasetPtr output_dataset{PGM_create_dataset_mutable(hl, "sym_output", 0, 1)};
+            PGM_dataset_mutable_add_buffer(hl, output_dataset.get(), "node", 1, 1, nullptr, sym_node_output.data());
+            // run power flow
+            OptionPtr options{PGM_create_options(hl)};
+            PGM_calculate(hl, model.get(), options.get(), output_dataset.get(), nullptr);
+            // compare output
+            CHECK(sym_node_output[0].id == node[0].id);
+            CHECK(sym_node_output[0].u_pu == doctest::Approx{1.0});
```
Comment (mgovers) [inline tests/c_api_tests/test_serialization.cpp:163]:
```suggestion
```

**Review by nitbharambe — APPROVED** [review]

**Review by Jerry-Jinfeng-Guo — APPROVED** [review]
The PR title is not the most descriptive one unfortunately ;)

**Pull request merged**